File: frontend/src/engine/capabilities/normalizeDates.ts
/**
 * normalizeDates – Standardize date formats in a column to a consistent format.
 *
 * Handles multiple input formats:
 *   dd/mm/yyyy, mm/dd/yyyy, yyyy-mm-dd, d-MMM-yy, Excel serial numbers, etc.
 *
 * Excel serial numbers are converted using the epoch offset (25569 days from
 * 1900-01-01 to 1970-01-01) and 86400000 ms per day.
 */

import { CapabilityMeta, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "normalizeDates",
  description: "Standardize date formats in a column to a consistent output format",
  mutates: true,
  affectsFormatting: false,
};

/** Months abbreviated for d-MMM-yy parsing */
const MONTH_ABBR: Record<string, number> = {
  jan: 0, feb: 1, mar: 2, apr: 3, may: 4, jun: 5,
  jul: 6, aug: 7, sep: 8, oct: 9, nov: 10, dec: 11,
};

const EXCEL_EPOCH = 25569;
const MS_PER_DAY = 86400000;

function tryParseDate(val: string | number | boolean | null): Date | null {
  if (val === null || val === "") return null;

  // Excel serial number
  if (typeof val === "number") {
    if (val > 1 && val < 2958466) {
      // Valid Excel date serial range
      return new Date((val - EXCEL_EPOCH) * MS_PER_DAY);
    }
    return null;
  }

  const s = String(val).trim();

  // yyyy-mm-dd or yyyy/mm/dd
  const isoMatch = s.match(/^(\d{4})[/-](\d{1,2})[/-](\d{1,2})$/);
  if (isoMatch) {
    return new Date(+isoMatch[1], +isoMatch[2] - 1, +isoMatch[3]);
  }

  // dd/mm/yyyy or dd-mm-yyyy
  const dmy = s.match(/^(\d{1,2})[/.-](\d{1,2})[/.-](\d{4})$/);
  if (dmy) {
    const day = +dmy[1];
    const month = +dmy[2];
    // Heuristic: if first number > 12, it must be day
    if (day > 12) {
      return new Date(+dmy[3], month - 1, day);
    }
    // If second number > 12, it must be day (mm/dd/yyyy)
    if (month > 12) {
      return new Date(+dmy[3], day - 1, month);
    }
    // Ambiguous — assume dd/mm/yyyy
    return new Date(+dmy[3], month - 1, day);
  }

  // d-MMM-yy or d-MMM-yyyy (e.g. 5-Jan-23, 15-Mar-2023)
  const mmmMatch = s.match(/^(\d{1,2})[/-]([A-Za-z]{3})[/-](\d{2,4})$/);
  if (mmmMatch) {
    const mon = MONTH_ABBR[mmmMatch[2].toLowerCase()];
    if (mon !== undefined) {
      let year = +mmmMatch[3];
      if (year < 100) year += year < 50 ? 2000 : 1900;
      return new Date(year, mon, +mmmMatch[1]);
    }
  }

  // Fallback: built-in Date.parse
  const fallback = Date.parse(s);
  if (!isNaN(fallback)) return new Date(fallback);

  return null;
}

function formatDate(d: Date, fmt: string): string {
  const pad = (n: number) => n.toString().padStart(2, "0");
  const yyyy = d.getFullYear().toString();
  const mm = pad(d.getMonth() + 1);
  const dd = pad(d.getDate());
  return fmt
    .replace(/yyyy/gi, yyyy)
    .replace(/mm/g, mm)
    .replace(/dd/g, dd);
}

async function handler(
  context: Excel.RequestContext,
  params: any,
  options: ExecutionOptions,
): Promise<StepResult> {
  const { range, outputFormat, inputFormat: _inputFormat } = params as {
    range: string;
    outputFormat: string;
    inputFormat?: string;
  };

  if (options.dryRun) {
    return {
      stepId: "",
      status: "success",
      message: `Would normalize dates in ${range} to format "${outputFormat}"`,
    };
  }

  options.onProgress?.("Reading range...");
  const rawRng = resolveRange(context, range);
  // Clip to used range — full-column refs like "A:A" would load ~1M rows
  let rng: Excel.Range;
  try {
    rng = rawRng.getUsedRange(false);
  } catch {
    rng = rawRng;
  }
  rng.load("values");
  await context.sync();

  const vals = (rng.values ?? []) as (string | number | boolean | null)[][];
  if (!vals.length) return { stepId: "", status: "success", message: "No data found." };

  options.onProgress?.("Normalizing dates...");
  let normalized = 0;
  const total = vals.reduce((sum, row) => sum + row.length, 0);
  const out: (string | number | boolean | null)[][] = vals.map((r) => [...r]);

  for (let r = 0; r < out.length; r++) {
    for (let c = 0; c < out[r].length; c++) {
      const parsed = tryParseDate(out[r][c]);
      if (parsed) {
        out[r][c] = formatDate(parsed, outputFormat);
        normalized++;
      }
    }
  }

  rng.values = out as any;
  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Normalized ${normalized}/${total} dates to "${outputFormat}"`,
    outputs: { range },
  };
}

registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/normalize_dates.py
"""normalizeDates — standardise date formats in a column.

Ported from `frontend/src/engine/capabilities/normalizeDates.ts`. Handles:
- Excel serial numbers (floats in a sensible range)
- yyyy-mm-dd / yyyy/mm/dd
- dd/mm/yyyy / mm/dd/yyyy with a heuristic disambiguation
- d-MMM-yy / d-MMM-yyyy
- datetime values coming directly from xlwings
- fallback to dateutil.parser when installed
"""

from __future__ import annotations

import re
from datetime import datetime
from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range

try:
    from dateutil import parser as _dateutil_parser  # type: ignore
except Exception:  # noqa: BLE001
    _dateutil_parser = None  # type: ignore


_MONTH_ABBR = {
    "jan": 1, "feb": 2, "mar": 3, "apr": 4, "may": 5, "jun": 6,
    "jul": 7, "aug": 8, "sep": 9, "oct": 10, "nov": 11, "dec": 12,
}

_EXCEL_EPOCH_DAYS = 25569

_ISO_RE = re.compile(r"^(\d{4})[/-](\d{1,2})[/-](\d{1,2})$")
_DMY_RE = re.compile(r"^(\d{1,2})[/.\-](\d{1,2})[/.\-](\d{4})$")
_MMM_RE = re.compile(r"^(\d{1,2})[/-]([A-Za-z]{3})[/-](\d{2,4})$")


def _to_2d(values: Any) -> list[list[Any]]:
    if values is None:
        return []
    if not isinstance(values, list):
        return [[values]]
    if not values:
        return []
    if not isinstance(values[0], list):
        return [values]
    return values


def _try_parse_date(val: Any) -> datetime | None:
    if val is None or val == "":
        return None
    if isinstance(val, datetime):
        return val

    if isinstance(val, (int, float)) and not isinstance(val, bool):
        if 1 < val < 2958466:
            return datetime.utcfromtimestamp((val - _EXCEL_EPOCH_DAYS) * 86400)
        return None

    s = str(val).strip()

    m = _ISO_RE.match(s)
    if m:
        try:
            return datetime(int(m.group(1)), int(m.group(2)), int(m.group(3)))
        except ValueError:
            return None

    m = _DMY_RE.match(s)
    if m:
        a, b, y = int(m.group(1)), int(m.group(2)), int(m.group(3))
        try:
            if a > 12:
                return datetime(y, b, a)
            if b > 12:
                return datetime(y, a, b)
            # Ambiguous → assume dd/mm/yyyy (TS uses same default)
            return datetime(y, b, a)
        except ValueError:
            return None

    m = _MMM_RE.match(s)
    if m:
        mon = _MONTH_ABBR.get(m.group(2).lower())
        if mon is not None:
            year = int(m.group(3))
            if year < 100:
                year += 2000 if year < 50 else 1900
            try:
                return datetime(year, mon, int(m.group(1)))
            except ValueError:
                return None

    if _dateutil_parser is not None:
        try:
            return _dateutil_parser.parse(s)
        except (ValueError, OverflowError, TypeError):
            return None

    return None


def _format_date(d: datetime, fmt: str) -> str:
    yyyy = f"{d.year:04d}"
    mm = f"{d.month:02d}"
    dd = f"{d.day:02d}"
    out = re.sub(r"yyyy", yyyy, fmt, flags=re.IGNORECASE)
    out = out.replace("mm", mm)
    out = out.replace("dd", dd)
    return out


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    address = params.get("range")
    output_format = params.get("outputFormat")

    if not address or not output_format:
        return {"status": "error", "message": "normalizeDates requires 'range' and 'outputFormat'."}

    rng = resolve_range(ctx.workbook_handle, address)

    if ctx.dry_run:
        return {
            "status": "preview",
            "message": f"Would normalize dates in {rng.address} to '{output_format}'.",
        }

    vals = _to_2d(rng.value)
    if not vals:
        return {"status": "success", "message": "No data found.", "outputs": {"range": rng.address}}

    out: list[list[Any]] = [list(row) for row in vals]
    total = sum(len(r) for r in out)
    normalized = 0

    for r in range(len(out)):
        for c in range(len(out[r])):
            parsed = _try_parse_date(out[r][c])
            if parsed is not None:
                out[r][c] = _format_date(parsed, output_format)
                normalized += 1

    try:
        rng.value = out
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"Write failed: {exc}", "error": str(exc)}

    return {
        "status": "success",
        "message": f"Normalized {normalized}/{total} dates to '{output_format}' in {rng.address}.",
        "outputs": {"range": rng.address},
    }


registry.register("normalizeDates", handler, mutates=True)
